modules/docs/arrow-docs/docs/docs/arrow/typeclasses/traverse/README.md

```diff
@@ -11,16 +11,491 @@ redirect_from:
 {:.intermediate}
 intermediate
 
-TODO. Meanwhile you can find a short description in the [intro to typeclasses]({{ '/docs/typeclasses/intro/' | relative_url }}).
+In functional programming it is very common to encode "behaviors" as data types - common behaviors include `Option` for possibly missing values, `Either` and `Validated` for possible errors, and [`Promise`]({{ '/docs/effects/promise/' | relative_url }}) for asynchronous computations, which has the same purpose as a `Future`, only pure and lazy.
 
+These behaviors tend to show up in functions working on a single piece of data - for instance parsing a single `String` into an `Int`, validating a login, or asynchronously fetching website information for a user.
+
+```kotlin:ank
+import arrow.core.Either
+import arrow.core.Option
+import arrow.fx.ForIO
+import arrow.fx.Promise
+
+interface Profile
+interface User
+
+fun userInfo(u: User): Promise<ForIO, Profile> = TODO()
+```
+
+Each function asks only for the data it actually needs; in the case of `userInfo`, a single `User`. We certainly could write one that takes a `List<User>` and fetch profile for all of them, would be a bit strange. If we just wanted to fetch the profile of just one user, we would either have to wrap it in a `List` or write a separate function that takes in a single user anyways. More fundamentally, functional programming is about building lots of small, independent pieces and composing them to make larger and larger pieces - does this hold true in this case?
+
+Given just `(User) -> Promise<ForIO, Profile>`, what should we do if we want to fetch profiles for a `List<User>`? We could try familiar combinators like map.
+
+```kotlin:ank
+fun profilesFor(users: List<User>): List<Promise<ForIO, Profile>> = users.map(::userInfo)
+```
+
+Note the return type `List<Promise<ForIO, Profile>>`. This makes sense given the type signatures, but seems unwieldy. We now have a list of asynchronous values, and to work with those values we must then use the combinators on `Promise` for every single one. It would be nicer instead if we could get the aggregate result in a single `Promise`, say a `Promise<ForIO, List<Profile>>`.
+
+There exists a much more generalized form that would allow us to parse a `List<String>` or validate credentials for a `List<User>`.
+
+Enter `Traverse`.
+
+### The Typeclass
+
+At center stage of Traverse is the `traverse` method.
+
+```kotlin
+import arrow.Kind
+import arrow.typeclasses.Applicative
+import arrow.typeclasses.Foldable
+import arrow.typeclasses.Functor
+
+interface Traverse<F> : Functor<F>, Foldable<F> {
+  fun <G, A, B> Kind<F, A>.traverse(AP: Applicative<G>, f: (A) -> Kind<G, B>): Kind<G, Kind<F, B>>
+}
+```
+
+In our above example, `F` is `List`, and `G` is `Option`, `Either`, or `Promise`. For the profile example, traverse says given a `List<User>` and a function `(User) -> Promise<ForIO, Profile>`, it can give you a `Promise<ForIO, List<Profile>>`.
```

Comment:
So `F` could be named a `container`, am I right?

Reply:
I'd love to have readable type parameters instead of `G`, `A`, `B` which convey no information in themselves.

Reply:
They have to be polymorphic :) @adam-arold
Yes @rcd27 , `F` can be seen as a "container", to be precise it is a parameter for type constructors, e.g.: containers like List, Either..

Reply:
@i-walker what I meant is that you could use `CONTAINER` instead of `T` in the example above. The point is that the generic type parameter would convey what it stands for not just an arbitrary letter like `F`.